File: scripts/wire.lua
-- Wire structure placement + connection-aware autotile (#359, power
-- epic #357).
--
-- Wire is a structure piece ("wire" pack, "wire" kind, no edge — see
-- data/structure_packs/wire.yaml) that occupies the tile top like a
-- floor (Structure.Types.SWire). Its rendered variant is DERIVED from
-- which of the tile's 4 cardinal neighbours also carry wire, so a laid
-- run reads as one continuous line. Placing a wire tile re-derives its
-- OWN shape and re-caps every WIRED neighbour (their shape may have
-- changed too) — mirrors scripts/structures.lua's post/wall
-- cap-recompute (recapTileCorner).
--
-- No M.clear: there is no demolish/removal flow for any structure piece
-- yet (#359's "removal/rerouting" is explicitly deferred), and
-- structure.clear's async apply means an immediate neighbour re-cap
-- would still see the piece via the authoritative overlay (the staging
-- cache only fast-paths ADDS, not clears) — add it once a real caller
-- needs it, matching against that behaviour rather than guessing at it.
--
-- Placement always targets the ACTIVE world: the only callers are the
-- interactive build-tool picker and the construct_job build AI, neither
-- of which needs the worldId/baseZ escape hatches structures.lua's
-- location-stamping builders carry.

local M = {}

local PACK_PATH = "data/structure_packs/wire.yaml"

local packCache = nil
local function packDef()
    if packCache then return packCache end
    packCache = engine.loadYaml(PACK_PATH)
    if not packCache then
        engine.logWarn("wire: failed to load pack '" .. PACK_PATH .. "'")
    end
    return packCache
end

local handleCache = nil
local function handles()
    if handleCache then return handleCache end
    local pack = packDef()
    if not pack then return { conn = {}, connPath = {} } end
    local conn, connPath = {}, {}
    for name, path in pairs(pack.connections or {}) do
        conn[name] = engine.loadTexture(path)
        connPath[name] = path
    end
    handleCache = { conn = conn, connPath = connPath,
                    face = engine.loadTexture(pack.facemap),
                    facePath = pack.facemap }
    return handleCache
end

-- Cardinal neighbour offsets, matching the engine's slope-bit convention
-- (World.Slope: bit0=N(gy-1) bit1=E(gx+1) bit2=S(gy+1) bit3=W(gx-1)).
local NEIGHBOR_OFFSETS = { {0, -1}, {1, 0}, {0, 1}, {-1, 0} }

-- Which neighbours a wire at (gx,gy) connects to. The engine owns this
-- (#1842) so the construction render pass and the placer cannot drift:
-- one canonical, seam-aware (#1175) lookup of the same placed + staged
-- wire `structure.hasAt` reads. The designation-aware variant is the
-- render pass's — `structure.wireNeighbors(gx, gy, page, true)` — and
-- placement must NOT use it, or laying a run would recap neighbours
-- against wire that is not there yet.
local function neighborsAt(gx, gy)
    local n = structure.wireNeighbors(gx, gy)
    return n.n, n.e, n.s, n.w
end

-- Which connection-shape a tile's 4-neighbour wire presence maps to.
-- Also engine-side now (#1842): ONE sixteen-way rule, shared with the
-- render pass, rather than a second table to keep in step. Corner/tee
-- names follow the wall-edge vocabulary (ne/nw/se/sw); a tee is named by
-- its MISSING side (e.g. tee_n = connected E+S+W, the open side facing N).
local function shapeFor(n, e, s, w)
    return structure.wireShape(n, e, s, w)
end

-- Every connection variant the pack must declare, in the shape rule's
-- own vocabulary.
local WIRE_SHAPES = {
    "isolated", "end_n", "end_e", "end_s", "end_w",
    "straight_ns", "straight_ew",
    "corner_ne", "corner_nw", "corner_se", "corner_sw",
    "tee_n", "tee_e", "tee_s", "tee_w", "cross",
}

-- Declare the wire pack's per-shape art for UNPLACED pieces (#1842), so
-- the construction render pass can resolve what a wire designation would
-- be BUILT with without calling into Lua. One call per session,
-- registered all-or-nothing engine-side; a shape the pack never declared
-- is OMITTED rather than sent with a nil path, so the engine's refusal
-- names the missing connection instead of rejecting the payload as
-- unreadable. It reports the failure itself — nothing warns here.
--
-- The pack NAME is the literal "wire": that is what a wire designation
-- carries (scripts/build_tool.lua's `pack = "wire"`), and the catalogue
-- is keyed by the designation's own spelling.
local registeredArt = false
function M.registerPackArt()
    if registeredArt or not structure.registerPackArt then return end
    local pack = packDef()
    if not pack then return end
    registeredArt = true
    local h = handles()
    local b = pack.build and pack.build.wire
    local art = {}
    for _, name in ipairs(WIRE_SHAPES) do
        if h.connPath[name] and h.facePath then
            art[#art + 1] = { kind = "wire", shape = name,
                              texture = h.connPath[name], texHandle = h.conn[name],
                              facemap = h.facePath, faceHandle = h.face }
        end
    end
    -- #1844: a buildable kind states its exact cost, which is what the
    -- engine charges and what a refund receipt records.
    local buildable = (b ~= nil) and (b.build_work ~= nil)
                      and (b.materials ~= nil)
    local kindEntry = { kind = "wire", buildable = buildable }
    if buildable then
        kindEntry.build_work = b.build_work
        kindEntry.materials = b.materials
    end
    structure.registerPackArt{
        pack  = "wire",
        kinds = { kindEntry },
        art   = art,
    }
end

-- (Re)place the wire piece at (gx,gy) with the shape its CURRENT
-- neighbours dictate. Does not touch neighbours — M.place re-caps those
-- separately below. Returns true/false (+ a reason on false) so
-- M.place can report the outcome of the tile the player actually
-- committed (a re-cap of an already-wired neighbour is incidental, not
-- the commit itself). Propagates structure.place's own result — it
-- returns false without placing anything when there's no active world
-- or the target chunk is unloaded (review round 7: this was previously
-- discarded, so a failed placement still recorded "accepted").
local function placeSelf(gx, gy)
    local h = handles()
    local shape = shapeFor(neighborsAt(gx, gy))
    local tex, path = h.conn[shape], h.connPath[shape]
    if not tex or not path then
        return false, "wire connection texture pack failed to load"
    end
    local z = (world.getTerrainAt(gx, gy) or 0) + 1
    local placed = structure.place(gx, gy, "wire", tex, h.face, z, path, h.facePath)
    if not placed then
        return false, "no active world or unloaded target chunk"
    end
    return true
end

-- Place wire at (gx,gy), deriving its shape from current neighbours, and
-- re-cap any wired neighbours whose own shape now includes this tile.
-- Order-independent, like structures.lua's wall/post placement.
function M.place(gx, gy)
    -- placeSelf's success path returns a single value (true, no second
    -- return), so failReason is already nil whenever ok is true —
    -- `reason = failReason` is correct as-is. Do NOT write this as
    -- `ok and nil or failReason`/`ok and nil or "some constant"`: with a
    -- non-nil constant fallback that idiom always selects the constant
    -- regardless of ok, because `ok and nil` collapses to a falsy value
    -- either way (review round 7 — a successful placement recorded a
    -- failure reason).
    local ok, failReason = placeSelf(gx, gy)
    debug.recordOutcome{
        kind = "wire.place", outcome = ok and "accepted" or "rejected",
        where = { x = gx, y = gy },
        reason = failReason,
    }
    for _, o in ipairs(NEIGHBOR_OFFSETS) do
        local nx, ny = gx + o[1], gy + o[2]
        if structure.hasAt(nx, ny, "wire") then
            placeSelf(nx, ny)
        end
    end
    return ok
end

return M

File: src/Structure/Wire.hs
{-# LANGUAGE Strict #-}
-- | The wire autotile RULE (#359, exposed engine-side by #1842): which
--   of the "wire" pack's sixteen connection variants a tile draws, given
--   which of its four cardinal neighbours also carry wire.
--
--   Until #1842 this lived only in @scripts/wire.lua@'s @shapeFor@, which
--   is fine while the only consumer is the placer — but the render pass
--   has to answer the same question for a wire that is merely
--   DESIGNATED, and the world render thread cannot call into Lua. Rather
--   than grow a second implementation that would drift, the rule moves
--   here and @scripts/wire.lua@ calls it through @structure.wireShape@:
--   one function, one vocabulary, one set of sixteen names.
--
--   The names are the pack YAML's own connection keys
--   (@data/structure_packs/wire.yaml@), so 'wireShapeName' is what
--   indexes the art and must not be "tidied": a tee is named by its
--   MISSING side (@tee_n@ = connected E+S+W, the open side facing N),
--   which reads backwards until you notice it names the gap, and the
--   corner/tee vocabulary is the wall edges' (@ne@\/@nw@\/@se@\/@sw@).
module Structure.Wire
    ( WireShape(..)
    , WireNeighbors(..)
    , wireShapeFor
    , wireShapeName
    , wireShapeFromName
    , allWireShapes
    , allWireNeighbors
    ) where

import UPrelude

-- | One of the sixteen connection variants. Enumerated (not a
--   'Text') so a resolution cannot invent a name the pack has no art
--   for, and 'Bounded'\/'Enum' so a test can sweep the whole space.
data WireShape
    = WireIsolated
    | WireEndN | WireEndE | WireEndS | WireEndW
    | WireStraightNS | WireStraightEW
    | WireCornerNE | WireCornerNW | WireCornerSE | WireCornerSW
    | WireTeeN | WireTeeE | WireTeeS | WireTeeW
    | WireCross
    deriving (Show, Eq, Ord, Enum, Bounded)

-- | Which cardinal neighbours of a tile carry wire, in the engine's own
--   slope-bit order (@World.Slope@: N = gy-1, E = gx+1, S = gy+1,
--   W = gx-1) — the order @scripts/wire.lua@'s @NEIGHBOR_OFFSETS@ walks.
data WireNeighbors = WireNeighbors
    { wnNorth ∷ !Bool
    , wnEast  ∷ !Bool
    , wnSouth ∷ !Bool
    , wnWest  ∷ !Bool
    } deriving (Show, Eq, Ord)

-- | The shape a tile with these neighbours draws. A transliteration of
--   @scripts/wire.lua@'s @shapeFor@, branch for branch, including its
--   count-first structure: the two-neighbour case is the only one where
--   WHICH pair matters, and the straights must be tested before the
--   corners.
wireShapeFor ∷ WireNeighbors → WireShape
wireShapeFor (WireNeighbors n e s w)
    | count ≡ 0 = WireIsolated
    | count ≡ 4 = WireCross
    | count ≡ 1 = if n then WireEndN
             else if e then WireEndE
             else if s then WireEndS
             else           WireEndW
    | count ≡ 3 = if not n then WireTeeN
             else if not e then WireTeeE
             else if not s then WireTeeS
             else               WireTeeW
    | n ∧ s     = WireStraightNS
    | e ∧ w     = WireStraightEW
    | n ∧ e     = WireCornerNE
    | n ∧ w     = WireCornerNW
    | s ∧ e     = WireCornerSE
    | otherwise = WireCornerSW
  where
    count = length (filter id [n, e, s, w]) ∷ Int

-- | The pack YAML's connection key for a shape. The ONE spelling — the
--   catalogue is keyed by it and the art is looked up by it.
wireShapeName ∷ WireShape → Text
wireShapeName s = case s of
    WireIsolated    → "isolated"
    WireEndN        → "end_n"
    WireEndE        → "end_e"
    WireEndS        → "end_s"
    WireEndW        → "end_w"
    WireStraightNS  → "straight_ns"
    WireStraightEW  → "straight_ew"
    WireCornerNE    → "corner_ne"
    WireCornerNW    → "corner_nw"
    WireCornerSE    → "corner_se"
    WireCornerSW    → "corner_sw"
    WireTeeN        → "tee_n"
    WireTeeE        → "tee_e"
    WireTeeS        → "tee_s"
    WireTeeW        → "tee_w"
    WireCross       → "cross"

-- | Parse a connection key. Exact, case-sensitively — the pack YAML's
--   keys are data, and a lenient parse would silently accept art the
--   placer would never ask for.
wireShapeFromName ∷ Text → Maybe WireShape
wireShapeFromName t =
    lookup t [ (wireShapeName s, s) | s ← allWireShapes ]

allWireShapes ∷ [WireShape]
allWireShapes = [minBound .. maxBound]

-- | The full sixteen-element neighbour space, so a sweep cannot miss a
--   combination.
allWireNeighbors ∷ [WireNeighbors]
allWireNeighbors =
    [ WireNeighbors n e s w
    | n ← [False, True], e ← [False, True]
    , s ← [False, True], w ← [False, True] ]
